File: Top_100_Liked_Questions/70._Climbing_Stairs/solution.py
class Solution:
    def climbStairs2(self, n: int) -> int:
        if n == 0:
            return 1
        elif n < 0:
            return 0
        
        # you can take 1 step and have climbStairs(n-1) number of ways
        # or you can take 2 steps and have climbStairs(n-2) number of ways
        
        return self.climbStairs(n-1) + self.climbStairs(n-2)
    
    
    def climbStairs(self, n: int) -> int:
        if not n:
            return 0
        
        # the above is equivalent to the Fibonacci sequence; n_2 = n_1 + n_0
        # for n times, update n_1 <- n_1 + n_0 and n_0 <- n_1
        i, j = 1, 0
        for _ in range(n):
            i, j = i + j, i
        
        return i

File: Top_100_Liked_Questions/70._Climbing_Stairs/solution.java
class Solution {
    public int climbStairs(int n) {
        // take 1 step and left with (n-1) steps
        // take 2 steps and left with (n-2) steps
        // fibonacci: f(n) = f(n-1) + f(n-2)

        int i=1;
        int j=0;
        int temp;

        for (int k=0; k<n; k++) {
            temp = i;
            i = i + j;
            j = temp;
        }

        return i;
    }
}